README.md
<h2><b>parent_child_checkbox</b></h2>

<h3>Parent Child Checkbox is a type of checkbox where we can establish hierarchy in Checkboxes.</h3>
You can create a Parent Checkbox which has multiple checkboxes as their children! The value of parent checkbox depends on the value of children checkboxes and vice-versa.<br>
If we clicked on Parent Checkbox, all children checkboxes will be selected too. <br><br>

![alt text](https://miro.medium.com/max/1200/1*uq4jGrYXhZMEcXOv04mcoA.gif)

<h3>Properties available: </h3>
<table>
  <th>Parameter Name</th>
  <th>Parameter Type</th>
  <th>Description</th>
  <tr>
    <td>parent</td>
    <td>Text</td>
    <td>Text Widget to specify the Parent checkbox</td>
  </tr>
  <tr>
    <td>children</td>
    <td>List&ltText&gt</td>
    <td>List&ltText&gt Widgets to specify the Children checkboxes</td>
  </tr>
  <tr>
    <td>parentCheckboxColor</td>
    <td>Color</td>
    <td>Color of Parent CheckBox<br>Defaults to [ThemeData.toggleableActiveColor]</td>
  </tr>
  <tr>
    <td>childrenCheckboxColor</td>
    <td>Text Widget</td>
    <td>Text Widget to specify the Parent checkbox</td>
  </tr>
    <tr>
    <td>parentCheckboxScale</td>
    <td>double</td>
    <td> Scale of the Parent CheckBox. Defaults to [1.0]</td>
  </tr>
<tr>
    <td>childrenCheckboxScale</td>
    <td>double</td>
    <td> Scale of the Children CheckBox. Defaults to [1.0]</td>
  </tr>
  <tr>
    <td>gap</td>
    <td>double</td>
    <td>Gap between the Parent and Children CheckBox. Defaults to [10.0]</td>
  </tr>
    <tr>
    <td>onCheckedChild</td>
    <td>void Function(int index)?</td>
    <td>Function that will be executed if a child will be selected</td>
  </tr>
 
  <tr>
    <td>onCheckedParent</td>
    <td>void Function(int index)?</td>
    <td>Function that will be executed if the parent will be selected</td>
  </tr>
  <tr>
    <td>isParentSelected</td>
    <td>Getter</td>
    <td>Getter to get whether particular parent is selected or not.<br><br>Example: {'Parent 1' : true, 'Parent 2' : false} where Parent 1 and Parent 2 will be 2 separate      parents if you are using multiple ParentChildCheckbox in your code.<br><br>
  Default value will be false for all specified parents
    </td>
  </tr>
  <tr>
    <td>selectedChildrens</td>
    <td>Getter</td>
    <td> Getter to get which childrens are selected for a particular parent.<br><br> Example: {'Parent 1' : ['Children 1.1','Children 1.2'], 'Parent 2' : []} where Parent 1 and Parent 2 will be 2 separate parents if you are using multiple ParentChildCheckbox in your code.<br><br> Default value is {'Parent 1' : [], 'Parent 2' : []}
    </td>
  </tr>
</table>

<h3>Example:</h3>
<br><br>

```
import 'dart:developer';
import 'package:flutter/material.dart';
import 'package:parent_child_checkbox/parent_child_checkbox.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      home: Scaffold(
        appBar: AppBar(title: Text('Parent Child Checkox')),
        body: Column(
          children: [
            ParentChildCheckbox(
              parent: Text('Parent 1'),
              children: [
                Text('Children 1.1'),
                Text('Children 1.2'),
                Text('Children 1.3'),
                Text('Children 1.4'),
              ],
              parentCheckboxColor: Colors.orange,
              childrenCheckboxColor: Colors.red,
            ),
            ParentChildCheckbox(
              parent: Text('Parent 2'),
              children: [
                Text('Children 2.1'),
                Text('Children 2.2'),
                Text('Children 2.3'),
                Text('Children 2.4'),
              ],
            ),
            ElevatedButton(
              child: Text('Get Data'),
              onPressed: () {
                log(ParentChildCheckbox.isParentSelected.toString());
                log(ParentChildCheckbox.selectedChildrens.toString());
              },
            ),
          ],
        ),
      ),
    );
  }
}


```


CHANGELOG.md
## 0.0.5

* Added onCheckedChild and onCheckedParent functionalities
* Added parentCheckboxScale and childrenCheckboxScale properties
* Added gap property 

## 0.0.4

* Updated deprecated APIs

## 0.0.3

* Formated code using flutter format .

## 0.0.2

* Updated Example to reflect new getter changes

## 0.0.1

* User can set activeColor of checkbox for both parent and children
* Added feature for having multiple ParentChildCheckbox
* Can get result of checkboxes using map `isParentSelected` and `selectedChildrenMap`
* Added null-safety



lib/parent_child_checkbox.dart
import 'package:flutter/material.dart';

///Main widget of Parent Child Checkbox.
///Parent Child Checkbox is a type of checkbox where we can establish hierarchy in Checkboxes.
class ParentChildCheckbox extends StatefulWidget {
  ///Default constructor of ParentChildCheckbox
  ParentChildCheckbox({
    required this.parent,
    required this.children,
    this.parentCheckboxColor,
    this.childrenCheckboxColor,
    this.parentCheckboxScale,
    this.childrenCheckboxScale,
    this.gap,
    this.onCheckedChild,
    this.onCheckedParent,
  });

  ///Text Widget to specify the Parent checkbox
  final Text? parent;

  ///List<Text> Widgets to specify the Children checkboxes
  final List<Text>? children;

  ///Color of Parent CheckBox
  ///
  /// By default the value of [CheckboxThemeData.fillColor]
  /// is used. If that is also null, then [ThemeData.disabledColor] is used in
  /// the disabled state, [ColorScheme.secondary] is used in the
  /// selected state, and [ThemeData.unselectedWidgetColor] is used in the
  /// default state.
  final Color? parentCheckboxColor;

  ///Color of Parent CheckBox
  ///
  /// By default the value of [CheckboxThemeData.fillColor]
  /// is used. If that is also null, then [ThemeData.disabledColor] is used in
  /// the disabled state, [ColorScheme.secondary] is used in the
  /// selected state, and [ThemeData.unselectedWidgetColor] is used in the
  /// default state.
  final Color? childrenCheckboxColor;

  ///Scale of the Parent CheckBox
  ///
  /// Defaults to [1.0]
  final double? parentCheckboxScale;

  ///Scale of the Children CheckBox
  ///
  /// Defaults to [1.0]
  final double? childrenCheckboxScale;

  ///Gap between the Parent and Children CheckBox
  ///
  /// Defaults to [10.0]
  final double? gap;

  ///Function that will be executed if a child will be selected
  ///
  final void Function(int index)? onCheckedChild;

  ///Function that will be executed if the parent will be selected
  ///
  final void Function()? onCheckedParent;

  /// Map which shows whether particular parent is selected or not.
  ///
  /// Example: {'Parent 1' : true, 'Parent 2' : false} where
  /// Parent 1 and Parent 2 will be 2 separate parents if you are using multiple ParentChildCheckbox in your code.
  ///
  /// Default value will be false for all specified parents
  static Map<String?, bool?> _isParentSelected = {};

  /// Getter to get whether particular parent is selected or not.
  ///
  /// Example: {'Parent 1' : true, 'Parent 2' : false} where
  /// Parent 1 and Parent 2 will be 2 separate parents if you are using multiple ParentChildCheckbox in your code.
  ///
  /// Default value will be false for all specified parents
  static get isParentSelected => _isParentSelected;

  /// Map which shows which childrens are selected for a particular parent.
  ///
  /// Example: {'Parent 1' : ['Children 1.1','Children 1.2'], 'Parent 2' : []} where
  /// Parent 1 and Parent 2 will be 2 separate parents if you are using multiple ParentChildCheckbox in your code.
  ///
  /// Default value is {'Parent 1' : [], 'Parent 2' : []}
  static Map<String?, List<String?>> _selectedChildrenMap = {};

  /// Getter to get which childrens are selected for a particular parent.
  ///
  /// Example: {'Parent 1' : ['Children 1.1','Children 1.2'], 'Parent 2' : []} where
  /// Parent 1 and Parent 2 will be 2 separate parents if you are using multiple ParentChildCheckbox in your code.
  ///
  /// Default value is {'Parent 1' : [], 'Parent 2' : []}
  static get selectedChildrens => _selectedChildrenMap;

  @override
  _ParentChildCheckboxState createState() => _ParentChildCheckboxState();
}

class _ParentChildCheckboxState extends State<ParentChildCheckbox> {
  ///Default parentValue set to false
  bool? _parentValue = false;

  ///List of childrenValue which depicts whether checkbox is clicked or not
  List<bool?> _childrenValue = [];

  @override
  void initState() {
    super.initState();
    _childrenValue = List.filled(widget.children!.length, false);
    ParentChildCheckbox._selectedChildrenMap.addAll({widget.parent!.data: []});
    ParentChildCheckbox._isParentSelected.addAll({widget.parent!.data: false});
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Row(
          children: [
            Transform.scale(
              scale: widget.parentCheckboxScale ?? 1.0,
              child: Checkbox(
                value: _parentValue,
                splashRadius: 0.0,
                activeColor: widget.parentCheckboxColor,
                onChanged: (value) => _parentCheckBoxClick(),
                tristate: true,
              ),
            ),
            SizedBox(
              width: 10.0,
            ),
            widget.parent!,
          ],
        ),
        SizedBox(
          height: widget.gap ?? 10.0,
        ),
        for (int i = 0; i < widget.children!.length; i++)
          Padding(
            padding: EdgeInsets.only(left: 25.0),
            child: Row(
              children: [
                Transform.scale(
                  scale: widget.childrenCheckboxScale ?? 1.0,
                  child: Checkbox(
                    splashRadius: 0.0,
                    activeColor: widget.childrenCheckboxColor,
                    value: _childrenValue[i],
                    onChanged: (value) => _childCheckBoxClick(i),
                  ),
                ),
                SizedBox(
                  width: 10.0,
                ),
                widget.children![i],
              ],
            ),
          ),
      ],
    );
  }

  ///onClick method when particular children of index i is clicked
  void _childCheckBoxClick(int i) {
    final Function(int index)? onCheckedChild = widget.onCheckedChild;
    if (onCheckedChild != null) {
      onCheckedChild(i);
    }
    setState(() {
      _childrenValue[i] = !_childrenValue[i]!;
      if (!_childrenValue[i]!) {
        ParentChildCheckbox._selectedChildrenMap.update(widget.parent!.data,
            (value) {
          value.removeWhere((element) => element == widget.children![i].data);
          return value;
        });
      } else {
        ParentChildCheckbox._selectedChildrenMap.update(widget.parent!.data,
            (value) {
          value.add(widget.children![i].data);
          return value;
        });
      }
      _parentCheckboxUpdate();
    });
  }

  ///onClick method when particular parent is clicked
  void _parentCheckBoxClick() {
    final Function()? onCheckedParent = widget.onCheckedParent;
    if (onCheckedParent != null) {
      onCheckedParent();
    }
    setState(() {
      if (_parentValue != null) {
        _parentValue = !_parentValue!;
        ParentChildCheckbox._isParentSelected
            .update(widget.parent!.data, (value) => _parentValue);
        ParentChildCheckbox._selectedChildrenMap.addAll({
          widget.parent!.data: [],
        });
        for (int i = 0; i < widget.children!.length; i++) {
          _childrenValue[i] = _parentValue;
          if (_parentValue!) {
            ParentChildCheckbox._selectedChildrenMap.update(widget.parent!.data,
                (value) {
              value.add(widget.children![i].data);
              return value;
            });
          }
        }
      } else {
        _parentValue = false;
        ParentChildCheckbox._isParentSelected
            .update(widget.parent!.data, (value) => _parentValue);
        ParentChildCheckbox._selectedChildrenMap
            .update(widget.parent!.data, (value) => []);
        for (int i = 0; i < widget.children!.length; i++)
          _childrenValue[i] = false;
      }
    });
  }

  ///Method to update the Parent Checkbox based on the status of Child checkbox
  void _parentCheckboxUpdate() {
    setState(() {
      if (_childrenValue.contains(false) && _childrenValue.contains(true)) {
        _parentValue = null;
        ParentChildCheckbox._isParentSelected
            .update(widget.parent!.data, (value) => false);
      } else {
        _parentValue = _childrenValue.first;
        ParentChildCheckbox._isParentSelected
            .update(widget.parent!.data, (value) => _childrenValue.first);
      }
    });
  }
}


example/lib/main.dart
import 'dart:developer';
import 'package:flutter/material.dart';
import 'package:parent_child_checkbox/parent_child_checkbox.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      home: Scaffold(
        appBar: AppBar(title: Text('Parent Child Checkox')),
        body: Column(
          children: [
            ParentChildCheckbox(
              parent: Text('Parent 1'),
              children: [
                Text('Children 1.1'),
                Text('Children 1.2'),
                Text('Children 1.3'),
                Text('Children 1.4'),
              ],
              parentCheckboxColor: Colors.orange,
              childrenCheckboxColor: Colors.red,
            ),
            ParentChildCheckbox(
              parent: Text('Parent 2'),
              children: [
                Text('Children 2.1'),
                Text('Children 2.2'),
                Text('Children 2.3'),
                Text('Children 2.4'),
              ],
            ),
            ElevatedButton(
              child: Text('Get Data'),
              onPressed: () {
                log(ParentChildCheckbox.isParentSelected.toString());
                log(ParentChildCheckbox.selectedChildrens.toString());
              },
            ),
          ],
        ),
      ),
    );
  }
}
